First pass at implementing conversion from XML to quasar.Data

jedesah opened the pull request:
I merged in master which is why the diff is bigger.

alissapajer committed:
Initial work on map func constant folding.

alissapajer committed:
Adds basic tests for constant folding.

alissapajer committed:
Adds more constant folding cases and a test.

alissapajer committed:
Merge remote-tracking branch 'upstream/master' into constant-fold

alissapajer committed:
Update comment.

alissapajer committed:
Remove unnecessary type annotations.

alissapajer committed:
Fix qscript tests to expect constant folding.

alissapajer committed:
Use orOriginal instead of repeatedly in constant folding.

alissapajer committed:
Remove duplication in constant folding impl.

wemrysi committed:
Add a companion object for MonadFsErr

jedesah committed:
Merge branch 'wemrysi-feature/1418-ml-main-module'

sellout committed:
Merge branch 'alissapajer-constant-fold'

jedesah committed:
Progress towards getting `trivial.test` to pass

- Apply `shiftRead` to QScript in planning
- Implement `idStatus` in `ShiftedReadPlanner`
- Add some XQuery functions
- Add a convenience method in `xquery.syntax`

jedesah committed:
Merge branch 'master' into feature/1372-ml-xml-data

jedesah committed:
First pass at implementing conversion from xml to Data

sellout committed:
Extract “frontend” project from “core”.

This project is conceptually something we want (so we can eventually
plug in multiple languages besides SQL²), but right now it definitely
has some extra bits in it. Including code that should become a separate
“sql” project.

Additionally, this introduces new packages with the names
`quasar.contrib.pathy` and `quasar.contrib.shapeless`. This is a
convention for where we put code that we would like to push upstream
– rather than hiding it in `quasar.fp` or just whereever we happen to
coding at the moment we need it.

wemrysi committed:
Disable sbt-header plugin for marklogic-validation until exclusion is available

sellout committed:
Don’t make doc warnings into errors, because scaladoc is very fragile.

wemrysi commented on marklogic/src/main/scala/quasar/physical/marklogic/xml/package.scala:
```diff
@@ -0,0 +1,85 @@
+    * }
+    */
+  def toData(elem: Elem): Data = {
+    Obj(ListMap(elem.label -> toDataImpl(elem.child.toList, elem.attributes.some)))
```
I don't think just using `label` will take namespace prefixes into account (though I could be mistaken given my complete lack of familiarity with `scala-xml`). i.e. if the element is `ejson:map`, `elem.label == "map"` and `elem.prefix == "ejson"`.

Also, and this is rather minor, but what do you think about just sticking with `Seq` and avoiding the `toList`s everywhere?

jedesah replied:
Not too surprisingly, you were right about `prefix + label`.

Because of the fact that `Data.Arr` wants a `List` and `Elem.chil` is a `Node*`. I think it's going to be hard to avoid many (if not all) of these `toList`. If you have something specific in mind, perhaps you could give it a quick try once this is merged.

wemrysi replied:
Ah, yeah, that's fine then.

wemrysi commented on marklogic/src/main/scala/quasar/physical/marklogic/xml/package.scala:
```diff
@@ -0,0 +1,85 @@
+      m.flatMap(attrToData).cata(
+        m => Obj(ListMap(
+          attributesKeyName -> m,
+          "_text"           -> Str(str)
```
How about making the `text` and `attributes` key names configurable, i.e. could be `toData(cfg: ToDataConfig, elem: Elem)` where config is like `ToDataConfig(attributesKeyName: String, textKeyName: String)`.

Then we could have a standard config that just uses `_attributes` and `_text`, respectively.

jedesah replied:
Done.

wemrysi commented on marklogic/src/test/scala/quasar/physical/marklogic/xml/ToDataSpec.scala:
```diff
@@ -0,0 +1,101 @@
+import quasar.Predef._
+import quasar.Data._
+
+class ToDataSpec extends quasar.Qspec {
```
Thanks for adding tests!

wemrysi replied:
Also, how about a test containing an empty element, not sure how we handle those.

jedesah replied:
I added tests, but did not implement yet because I am not sure about the expected result.

wemrysi commented on marklogic/src/main/scala/quasar/physical/marklogic/xml/package.scala:
```diff
@@ -0,0 +1,85 @@
+        Str(str)
+      )
+    case xs =>
+      val elementChildren = xs.collect { case a: Elem => a}.groupBy(_.label)
```
Same here, we probably want to group by `prefix + label`

wemrysi commented on marklogic/src/test/scala/quasar/physical/marklogic/xml/ToDataSpec.scala:
```diff
@@ -0,0 +1,101 @@
+import quasar.Predef._
+import quasar.Data._
+
+class ToDataSpec extends quasar.Qspec {
```
Should probably add one that uses namespaced element/attribute names to confirm or deny my suspicions about those.

alissapajer committed:
Adds changelog.

alissapajer committed:
Merge branch 'wemrysi-feature/untouched-headers'

sellout committed:
Merge branch 'sellout-modularize'

wemrysi committed:
Move MarkLogic planner effect to typeclass param and constraints to instance defns

jedesah committed:
Make `shiftedRead` a XQuery function

- Use `seqToArray` for convenience
- Add a `flakyTest` to the growing list

jedesah committed:
Merge branch 'master' into topic/marklogic_trivial_test

# Conflicts:
#	marklogic/src/main/scala/quasar/physical/marklogic/fs/queryfile.scala

wemrysi committed:
Merge remote-tracking branch 'upstream/master' into feature/ml-planner-constraints

jedesah committed:
Merge branch 'master' into topic/marklogic/xmlToData

jedesah committed:
Additional support for converting xml to data

- Add configurable xml "keywords"
- Add test case for elements with prefixes and modify implementation to satisfy spec
- Add pending test cases for empty elements

jedesah commented:
@wemrysi Thx for the feedback!

Let me know if you agree with the expected results for the two empty test cases and I will implement them to follow wtv spec we agree is best.

I am sure this doesn't cover everything, but hopefully, it's good enough to get started and we can extend it when we start hitting some real xml test cases.

jedesah committed:
Add `flakyTest`

drostron committed:
Merge branch 'wemrysi-feature/ml-planner-constraints'

jedesah committed:
Merge branch 'master' into topic/marklogic_trivial_test

# Conflicts:
#	marklogic/src/main/scala/quasar/physical/marklogic/qscript/ShiftedReadPlanner.scala

wemrysi commented:
Hm, do empty elements still contain a single `PCDATA` child with empty content? Or do they contain no children? The spec seems to imply they'd not contain any children, so maybe `null` would be a better choice for the value? Especially since `{"foo": ""}` is distinct from `{"foo": null}`. What do you think?

jedesah commented:
So interestingly, [this](http://convertjson.com/xml-to-json.htm) website does it exactly how I specified it (and I hadn't looked at any online stuff before writing those test cases). However [this one](http://www.utilities-online.info/xmltojson) uses `null`. But I don't like how that second website handles multiple empty elements with the same name, frankly, I would even consider it's handling of that case to be buggy which makes me think perhaps the authors of that tool didn't put too much thought behind handling corner cases.

[This SO question](http://stackoverflow.com/questions/774192/what-is-the-correct-way-to-represent-null-xml-elements) kind of also makes me lean towards the empty string. If there was another way in xml to represent the empty string I think I would prefer null in order for there not to be any ambiguity, but since it's the only way I know of.

In terms of implementing Data -> XML -> Data, how would we otherwise encode the empty string to return the empty string back? I mean we are going to have to encode ONE of them (empty string or `null`) differently and it seems somewhat more natural to handle all strings in the same way and do a special encoding for `null`.

What are you thoughts on these points?

wemrysi commented:
For XML representing a piece of `Data` we'll know we're looking at a particular encoding, so we can use attributes to hold the actual types, so there shouldn't be any ambiguity.

For general XML sounds like the empty string is the way to go, we can always change it later if we discover that to be a problematic representation.

wemrysi committed:
Merge branch 'jedesah-topic/marklogic_trivial_test'

jedesah committed:
Merge branch 'master' into topic/marklogic/xmlToData

jedesah committed:
Implement conversion of empty xml elements
